### ISSUE TITLE
test_double_compaction_by_cleanup_and_major_compactions: scylla ran OOM

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#14966]
Seen first in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/324/testReport/compaction_additional_test/TestCompactionAdditional/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split001___test_double_compaction_by_cleanup_and_major_compactions/
```
failed on teardown with "AssertionError: Critical errors found: [('node1', ['Aborting on shard 1.'])]
Other errors: [('node1', ['ERROR 2023-08-05 01:55:38,214 [shard 1] seastar_memory - Dumping seastar memory diagnostics', 'ERROR 2023-08-05 01:55:38,214 [shard 1] seastar - Failed to allocate 35856 bytes'])]"
```

The node log seems to show infinite recursion:

https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/324/artifact/logs-full.release.001/1691200540172_compaction_additional_test.py%3A%3ATestCompactionAdditional%3A%3Atest_double_compaction_by_cleanup_and_major_compactions/node1.log
```
INFO  2023-08-05 01:55:37,945 [shard 0] api - force_keyspace_cleanup: keyspace=ks tables={table{name=cf, id=0accb2c0-3333-11ee-b55e-ae9cf57101b8}}
INFO  2023-08-05 01:55:37,946 [shard 0] compaction_manager - Stopping 1 tasks for 0 ongoing compactions for table ks.cf compaction_group=0/1 due to user-triggered operation
INFO  2023-08-05 01:55:37,946 [shard 0] compaction_manager - The following SSTables require cleanup in this compaction: {/jenkins/workspace/scylla-master/dtest-release/scylla/.dtest/dtest-amxegjkg/test/node1/data/ks/cf-0accb2c0333311eeb55eae9cf57101b8/me-3g89_05cn_2edkg2rajchvejxkmw-big-Data.db:level=0:origin=memtable}
INFO  2023-08-05 01:55:37,946 [shard 0] compaction - [Cleanup ks.cf 2c964ba0-3333-11ee-b55e-ae9cf57101b8] Cleaning [/jenkins/workspace/scylla-master/dtest-release/scylla/.dtest/dtest-amxegjkg/test/node1/data/ks/cf-0accb2c0333311eeb55eae9cf57101b8/me-3g89_05cn_2edkg2rajchvejxkmw-big-Data.db:level=0:origin=memtable]
INFO  2023-08-05 01:55:37,946 [shard 1] compaction_manager - Stopping 1 tasks for 0 ongoing compactions for table ks.cf compaction_group=0/1 due to user-triggered operation
INFO  2023-08-05 01:55:37,946 [shard 1] compaction_manager - The following SSTables require cleanup in this compaction: {/jenkins/workspace/scylla-master/dtest-release/scylla/.dtest/dtest-amxegjkg/test/node1/data/ks/cf-0accb2c0333311eeb55eae9cf57101b8/me-3g89_05cn_2edkg273pgycfxvge0-big-Data.db:level=0:origin=memtable}
INFO  2023-08-05 01:55:37,946 [shard 1] compaction - [Cleanup ks.cf 2c964ba0-3333-11ee-907d-ae9df57101b8] Cleaning [/jenkins/workspace/scylla-master/dtest-release/scylla/.dtest/dtest-amxegjkg/test/node1/data/ks/cf-0accb2c0333311eeb55eae9cf57101b8/me-3g89_05cn_2edkg273pgycfxvge0-big-Data.db:level=0:origin=memtable]
INFO  2023-08-05 01:55:38,006 [shard 1] compaction - [Cleanup ks.cf 2c964ba0-3333-11ee-907d-ae9df57101b8] Cleaned 1 sstables to [/jenkins/workspace/scylla-master/dtest-release/scylla/.dtest/dtest-amxegjkg/test/node1/data/ks/cf-0accb2c0333311eeb55eae9cf57101b8/me-3g89_05cp_5mrdt273pgycfxvge0-big-Data.db:level=0]. 1MB to 1MB (~100% of original) in 59ms = 25MB/s. ~24960 total partitions merged to 24865.
INFO  2023-08-05 01:55:38,006 [shard 1] compaction_manager - Stopping 1 tasks for 0 ongoing compactions for table ks.cf compaction_group=0/1 due to user-triggered operation
...
INFO 2023-08-05 01:55:38,214 [shard 0] compaction_manager - Stopping 1 tasks for 0 ongoing compactions for table ks.cf compaction_group=0/1 due to user-triggered operation
ERROR 2023-08-05 01:55:38,214 [shard 1] seastar_memory - Dumping seastar memory diagnostics
Used memory:   462M
Free memory:   20K
Total memory:  462M
Hard failures: 1

LSA
  allocated: 256K
  used:      128K
  free:      128K

Cache:
  total: 256K
  used:  22K
  free:  234K

Memtables:
 total: 0B
 Regular:
  real dirty: 0B
  virt dirty: 0B
 System:
  real dirty: 0B
  virt dirty: 0B

Replica:
  Read Concurrency Semaphores:
    user: 0/100, 0B/9M, queued: 0
    streaming: 0/10, 0B/9M, queued: 0
    system: 0/10, 0B/9M, queued: 0
    compaction: 0/âˆž, 0B/âˆž
  Execution Stages:
    apply stage:
         Total: 0
  Tables - Ongoing Operations:
    Pending writes (top 10):
      0 Total (all)
    Pending reads (top 10):
      0 Total (all)
    Pending streams (top 10):
      0 Total (all)

Small pools:
objsz spansz usedobj memory unused wst%
    8     4K     370     8K     5K   63
   10     4K       1     8K     8K   99
   12     4K       0     0B     0B    0
   14     4K       0     0B     0B    0
   16     4K      5k    80K     5K    6
   32     4K      3k   100K     8K    7
   32     4K     17k   536K     4K    0
   32     4K      9k   292K     8K    2
   32     4K     23k   724K     6K    0
   48     4K      9k   424K     2K    0
   48     4K     941    56K    12K   21
   64     4K     22k     1M     4K    0
   64     4K     26k     2M     4K    0
   80     4K     12k   960K     4K    0
   96     4K     13k     1M     7K    0
  112     4K     469    64K    13K   19
  128     4K     963   148K    28K   18
  160     4K     18k     3M   320B    0
  192     4K     11k     2M    13K    0
  224     4K      2k   392K    28K    7
  256     4K     158    64K    25K   38
  320     8K     294   144K    52K   36
  384     8K     189    96K    25K   26
  448     4K     131   104K    46K   44
  512     4K     154   152K    75K   49
  640    16K      80   160K   110K   68
  768    16K     11k     8M    16K    0
  896     8K      37    72K    39K   54
 1024     4K     190   228K    38K   16
 1280    32K      55    96K    26K   27
 1536    32K       5   128K   120K   93
 1792    16K       7   144K   131K   91
 2048     8K     397   912K   118K   12
 2560    64K      85   384K   170K   44
 3072    64K       6   256K   237K   92
 3584    32K       3   288K   277K   96
 4096    16K      79   608K   292K   48
 5120   128K      17   384K   295K   76
 6144   128K       4   512K   486K   94
 7168    64K       2   576K   560K   97
 8192    32K     316     3M   512K   16
10240    64K       7   832K   760K   91
12288    64K      16   960K   768K   80
14336   128K       0     0B     0B    0
16384    64K      12     1M     1M   84

Page spans:
index  size  free  used spans
    0    4K    4K   13M    3k
    1    8K    0B    1M   153
    2   16K   16K    9M   557
    3   32K    0B    6M   186
    4   64K    0B  426M    7k
    5  128K    0B    3M    25
    6  256K    0B  512K     2
    7  512K    0B    0B     0
    8    1M    0B    0B     0
    9    2M    0B    0B     0
   10    4M    0B    4M     1
   11    8M    0B    0B     0
   12   16M    0B    0B     0
   13   32M    0B    0B     0
   14   64M    0B    0B     0
   15  128M    0B    0B     0
   16  256M    0B    0B     0
   17  512M    0B    0B     0
   18    1G    0B    0B     0
   19    2G    0B    0B     0
   20    4G    0B    0B     0
   21    8G    0B    0B     0
   22   16G    0B    0B     0
   23   32G    0B    0B     0
   24   64G    0B    0B     0
   25  128G    0B    0B     0
   26  256G    0B    0B     0
   27  512G    0B    0B     0
   28    1T    0B    0B     0
   29    2T    0B    0B     0
   30    4T    0B    0B     0
   31    8T    0B    0B     0

ERROR 2023-08-05 01:55:38,214 [shard 1] seastar - Failed to allocate 35856 bytes
Aborting on shard 1.
Backtrace:
  0x589d908
  0x58d2a72
  /jenkins/workspace/scylla-master/dtest-release/scylla/.ccm/scylla-repository/421a5ad55caa6ca0c88edd5e74fd1e47d9cb95cf/libreloc/libc.so.6+0x3db6f
  /jenkins/workspace/scylla-master/dtest-release/scylla/.ccm/scylla-repository/421a5ad55caa6ca0c88edd5e74fd1e47d9cb95cf/libreloc/libc.so.6+0x8e843
  /jenkins/workspace/scylla-master/dtest-release/scylla/.ccm/scylla-repository/421a5ad55caa6ca0c88edd5e74fd1e47d9cb95cf/libreloc/libc.so.6+0x3dabd
  /jenkins/workspace/scylla-master/dtest-release/scylla/.ccm/scylla-repository/421a5ad55caa6ca0c88edd5e74fd1e47d9cb95cf/libreloc/libc.so.6+0x2687e
  0x586ae67
  0x586ad3e
  0x586da01
  0x25d4ae4
  0x25bfcad
  0x25b2df3
  0x25b1546
  0x25b3cec
  0x25cc98f
  0x2658f41
  0x12f035a
  0x58aed3f
  0x58aff77
  0x58d3043
  0x587f52a
  /jenkins/workspace/scylla-master/dtest-release/scylla/.ccm/scylla-repository/421a5ad55caa6ca0c88edd5e74fd1e47d9cb95cf/libreloc/libc.so.6+0x8c906
  /jenkins/workspace/scylla-master/dtest-release/scylla/.ccm/scylla-repository/
```

**Comment by bhalevy** [comment]
@Deexie please look into this, it seems related to your latest changes around compaction tasks.

Decoded backtrace:
```
[Backtrace #0]
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/release/seastar/./seastar/include/seastar/util/backtrace.hh:64
 (inlined by) seastar::backtrace_buffer::append_backtrace() at ./build/release/seastar/./seastar/src/core/reactor.cc:818
 (inlined by) seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:848
seastar::print_with_backtrace(char const*, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:860
 (inlined by) seastar::sigabrt_action() at ./build/release/seastar/./seastar/src/core/reactor.cc:3976
 (inlined by) operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:3952
 (inlined by) __invoke at ./build/release/seastar/./seastar/src/core/reactor.cc:3948
/data/scylla-s3-reloc.cache/by-build-id/1c82e9a64c2243f24ce5b05ae125be2d8fcdc098/extracted/scylla/libreloc/libc.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=245240a31888ad5c11bbc55b18e02d87388f59a9, for GNU/Linux 3.2.0, not stripped

__GI___sigaction at :?
__pthread_kill_implementation at ??:?
__GI_raise at :?
__GI_abort at :?
seastar::memory::on_allocation_failure(unsigned long) at ./build/release/seastar/./seastar/src/core/memory.cc:1867
seastar::memory::allocate(unsigned long) at ./build/release/seastar/./seastar/src/core/memory.cc:1436
operator new(unsigned long) at ./build/release/seastar/./seastar/src/core/memory.cc:2122
seastar::chunked_fifo<seastar::internal::abortable_fifo<seastar::shared_future<std::optional<sstables::compaction_stats> >::shared_state::entry, seastar::shared_future<std::optional<sstables::compaction_stats> >::shared_state::entry_expiry>::entry, 128ul>::back_chunk_new() at ././seastar/include/seastar/core/chunked_fifo.hh:407
 (inlined by) seastar::chunked_fifo<seastar::internal::abortable_fifo<seastar::shared_future<std::optional<sstables::compaction_stats> >::shared_state::entry, seastar::shared_future<std::optional<sstables::compaction_stats> >::shared_state::entry_expiry>::entry, 128ul>::ensure_room_back() at ././seastar/include/seastar/core/chunked_fifo.hh:428
 (inlined by) void seastar::chunked_fifo<seastar::internal::abortable_fifo<seastar::shared_future<std::optional<sstables::compaction_stats> >::shared_state::entry, seastar::shared_future<std::optional<sstables::compaction_stats> >::shared_state::entry_expiry>::entry, 128ul>::emplace_back<>() at ././seastar/include/seastar/core/chunked_fifo.hh:464
 (inlined by) seastar::shared_future<std::optional<sstables::compaction_stats> >::shared_state::entry& seastar::internal::abortable_fifo<seastar::shared_future<std::optional<sstables::compaction_stats> >::shared_state::entry, seastar::shared_future<std::optional<sstables::compaction_stats> >::shared_state::entry_expiry>::emplace_back<>() at ././seastar/include/seastar/core/abortable_fifo.hh:214
 (inlined by) seastar::shared_future<std::optional<sstables::compaction_stats> >::shared_state::get_future(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >) at ././seastar/include/seastar/core/shared_future.hh:179
seastar::shared_future<std::optional<sstables::compaction_stats> >::get_future(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >) const at ././seastar/include/seastar/core/shared_future.hh:256
 (inlined by) compaction::compaction_task_executor::compaction_done() at ./compaction/compaction_manager.hh:555
 (inlined by) operator()<seastar::shared_ptr<compaction::compaction_task_executor> > at ./compaction/compaction_manager.cc:955
 (inlined by) seastar::future<void> seastar::futurize<seastar::future<void> >::invoke<compaction_manager::stop_tasks(std::vector<seastar::shared_ptr<compaction::compaction_task_executor>, std::allocator<seastar::shared_ptr<compaction::compaction_task_executor> > >, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0&, seastar::shared_ptr<compaction::compaction_task_executor>&>(compaction_manager::stop_tasks(std::vector<seastar::shared_ptr<compaction::compaction_task_executor>, std::allocator<seastar::shared_ptr<compaction::compaction_task_executor> > >, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0&, seastar::shared_ptr<compaction::compaction_task_executor>&) at ././seastar/include/seastar/core/future.hh:2006
 (inlined by) auto seastar::futurize_invoke<compaction_manager::stop_tasks(std::vector<seastar::shared_ptr<compaction::compaction_task_executor>, std::allocator<seastar::shared_ptr<compaction::compaction_task_executor> > >, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0&, seastar::shared_ptr<compaction::compaction_task_executor>&>(compaction_manager::stop_tasks(std::vector<seastar::shared_ptr<compaction::compaction_task_executor>, std::allocator<seastar::shared_ptr<compaction::compaction_task_executor> > >, seastar::basic_sstring<char, unsigned int, 15u, true>)::$_0&, seastar::shared_ptr<compaction::compaction_task_executor>&) at ././seastar/include/seastar/core/future.hh:2037
 (inlined by) parallel_for_each<__gnu_cxx::__normal_iterator<seastar::shared_ptr<compaction::compaction_task_executor> *, std::vector<seastar::shared_ptr<compaction::compaction_task_executor>, std::allocator<seastar::shared_ptr<compaction::compaction_task_executor> > > >, __gnu_cxx::__normal_iterator<seastar::shared_ptr<compaction::compaction_task_executor> *, std::vector<seastar::shared_ptr<compaction::compaction_task_executor>, std::allocator<seastar::shared_ptr<compaction::compaction_task_executor> > > >, (lambda at compaction/compaction_manager.cc:953:50)> at ././seastar/include/seastar/coroutine/parallel_for_each.hh:121
 (inlined by) parallel_for_each<std::vector<seastar::shared_ptr<compaction::compaction_task_executor>, std::allocator<seastar::shared_ptr<compaction::compaction_task_executor> > > &, (lambda at compaction/compaction_manager.cc:953:50)> at ././seastar/include/seastar/coroutine/parallel_for_each.hh:143
 (inlined by) compaction_manager::stop_tasks(std::vector<seastar::shared_ptr<compaction::compaction_task_executor>, std::allocator<seastar::shared_ptr<compaction::compaction_task_executor> > >, seastar::basic_sstring<char, unsigned int, 15u, true>) at ./compaction/compaction_manager.cc:953
compaction_manager::stop_ongoing_compactions(seastar::basic_sstring<char, unsigned int, 15u, true>, compaction::table_state*, std::optional<sstables::compaction_type>) at ./compaction/compaction_manager.cc:984
compaction_manager::stop_and_disable_compaction(compaction::table_state&) at ./compaction/compaction_manager.cc:645
compaction_manager::run_with_compaction_disabled(compaction::table_state&, std::function<seastar::future<void> ()>) at ./compaction/compaction_manager.cc:651
seastar::future<std::optional<sstables::compaction_stats> > compaction_manager::perform_task_on_all_files<compaction::cleanup_sstables_compaction_task_executor>(std::optional<tasks::task_info>, compaction::table_state&, sstables::compaction_type_options, seastar::lw_shared_ptr<std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > > const>, std::function<seastar::future<std::vector<seastar::lw_shared_ptr<sstables::sstable>, std::allocator<seastar::lw_shared_ptr<sstables::sstable> > > > ()>) at ./compaction/compaction_manager.cc:1486
compaction_manager::try_perform_cleanup(seastar::lw_shared_ptr<std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > > const>, compaction::table_state&, std::optional<tasks::task_info>) at ./compaction/compaction_manager.cc:1812
std::__n4861::coroutine_handle<seastar::internal::coroutine_traits_base<void>::promise_type>::resume() const at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/coroutine:240
 (inlined by) seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() at ././seastar/include/seastar/core/coroutine.hh:125
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at ./build/release/seastar/./seastar/src/core/reactor.cc:2597
 (inlined by) seastar::reactor::run_some_tasks() at ./build/release/seastar/./seastar/src/core/reactor.cc:3060
seastar::reactor::do_run() at ./build/release/seastar/./seastar/src/core/reactor.cc:3229
operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:4440
 (inlined by) void std::__invoke_impl<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0&>(std::__invoke_other, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/invoke.h:61
 (inlined by) std::enable_if<is_invocable_r_v<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0&>, void>::type std::__invoke_r<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0&>(seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/invoke.h:111
 (inlined by) std::_Function_handler<void (), seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0>::_M_invoke(std::_Any_data const&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:290
std::function<void ()>::operator()() const at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:591
 (inlined by) seastar::posix_thread::start_routine(void*) at ./build/release/seastar/./seastar/src/core/posix.cc:90
```

**Comment by denesb** [comment]
This seems like an important fix but it doesn't apply cleanly to 5.2. @Deexie please provide a bacport PR if you think this should be backported.

**Comment by Deexie** [comment]
> This seems like an important fix but it doesn't apply cleanly to 5.2. @Deexie please provide a bacport PR if you think this should be backported.

The change that introduced the bug isn't in 5.2

**Comment by denesb** [comment]
Ok, so there are no affected branches, removing backport candidate label.
